### PR TITLE
reconcile crd to desired state

### PR DESCRIPTION
**Pull request by ashleyschuett**
Signed-off-by: Ashley Schuett <aschuett@redhat.com>

**What this PR does / why we need it**:

This PR adds reconciliation of the CRD resources. If the spec or metadata of the cached CRD is different from the desired CRD we create a patch and update it. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
Reconcile CRD resources
```

### CODE REVIEW

File: pkg/virt-operator/resource/apply/crds_test.go [hunk 1]
```diff
@@ -198,4 +200,76 @@ var _ = Describe("Apply CRDs", func() {
...
+			targetStrategy: targetStrategy,
+			stores:         stores,
+			clientset:      clientset,
+			expectations:   expectations,
+		}
+
+		Expect(r.createOrUpdateCrd(cachedCRD)).To(BeNil())
+	})
+
+	It("should patch CRD on sync when it is not equal to the requried CRD", func() {
```
Comment (acardace) [inline pkg/virt-operator/resource/apply/crds_test.go:232]: 'requried' typo

File: pkg/virt-operator/resource/apply/crds.go [hunk 2]
```diff
@@ -47,42 +48,47 @@ func (r *Reconciler) createOrUpdateCrd(crd *extv1beta1.CustomResourceDefinition)
...
+	existing := cachedCrd.DeepCopy()
 
-		// Add Labels and Annotations Patches
-		labelAnnotationPatch, err := createLabelsAndAnnotationsPatch(&crd.ObjectMeta)
-		if err != nil {
-			return err
-		}
-		ops = append(ops, labelAnnotationPatch...)
+	resourcemerge.EnsureCustomResourceDefinitionV1Beta1(modified, existing, *crd)
+	// there was no change to metadata, the generation was right
```
Comment (acardace) [inline pkg/virt-operator/resource/apply/crds.go:55]: Maybe this comment is a copy and paste from somewhere else? nothing it's checking the generation, and the function is checking the whole CRD.

**Review by acardace — COMMENTED** [review]
just nits, patch looks good. Why didn't you use the same approach for webhooks (where you used the generation mechanisms)?

**Comment by ashleyschuett** [comment]
@acardace thanks for the review, i have update this PR. 

>just nits, patch looks good. Why didn't you use the same approach for webhooks (where you used the generation mechanisms)?

The difference with `webhooks` / `deployments` / `daemonsets` is that things can be added to their spec definition from their respective kubernetes controllers that control them. There can be fields add with defaulted values. By doing a equality comparison of their spec vs. our defined spec would not be equal, but the fields we care about could/would be equal. By tracking the generation we know that the resource was changed by someone that was not us and we can then just blindly update it to ensure it matches the definition that we want.

**Comment by acardace** [comment]
> The difference with `webhooks` / `deployments` / `daemonsets` is that things can be added to their spec definition from their respective kubernetes controllers that control them. There can be fields add with defaulted values. By doing a equality comparison of their spec vs. our defined spec would not be equal, but the fields we care about could/would be equal. By tracking the generation we know that the resource was changed by someone that was not us and we can then just blindly update it to ensure it matches the definition that we want.

Gotcha, thank you!

**Review by dhiller — APPROVED** [review]
/approve

**Pull request merged**